sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -149,18 +167,80 @@ private void Append(Data segment)
-                    ArraySegment<byte> arraySegment => arraySegment,
-                    _ => ReadOnlyMemory<byte>.Empty
-                };
+                var dataToAppend = Convert(segment);
 
                 var memory = _writer.GetMemory(dataToAppend.Length);
                 dataToAppend.CopyTo(memory);
                 _writer.Advance(dataToAppend.Length);
                 _segments.Add(memory.Slice(0, dataToAppend.Length));
             }
         }
+
+        /// <summary>
+        /// Eagerly copies the provided data segments into a single continuous buffer while still keeping around a list of the individual copied segments.
+        /// Important for the receive path in order to make sure the buffers managed by the underlying AMQP library can be released on dispose.
+        /// </summary>
+        private sealed class EagerCopyingMessageBodyWithManagedBuffer : MessageBody
+        {
+            private IList<ReadOnlyMemory<byte>> _segments;
+            private byte[] _continuousBuffer;
+            private ReadOnlyMemory<byte>? _writtenMemory;
+
+            internal EagerCopyingMessageBodyWithManagedBuffer(IEnumerable<Data> dataSegments)
+            {
+                Append(dataSegments);
+            }
+
+            protected override ReadOnlyMemory<byte> WrittenMemory => _writtenMemory ?? ReadOnlyMemory<byte>.Empty;
+
+            public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+            {
+                return _segments.GetEnumerator();
+            }
+
+            private void Append(IEnumerable<Data> dataSegments)
+            {
+                int length = 0;
+                List<ReadOnlyMemory<byte>> convertedSegments = null;
+                foreach (var dataSegment in dataSegments)
+                {
+                    convertedSegments ??= new List<ReadOnlyMemory<byte>>();
+                    ReadOnlyMemory<byte> readOnlyMemory = Convert(dataSegment);
+                    convertedSegments.Add(readOnlyMemory);
+                    length += readOnlyMemory.Length;
+                }
+
+                if (convertedSegments == null)
+                {
+                    return;
+                }
+
+                _segments ??= new List<ReadOnlyMemory<byte>>();
+                _continuousBuffer = ArrayPool<byte>.Shared.Rent(length);
+                Memory<byte> asMemory = _continuousBuffer.AsMemory();
+                int start = 0;
+                foreach (ReadOnlyMemory<byte> segment in convertedSegments)
+                {
+                    Memory<byte> destination = asMemory.Slice(start, segment.Length);
+                    segment.CopyTo(destination);
+                    _segments.Add(destination);
+                    start += segment.Length;
+                }
+
+                _writtenMemory = asMemory.Slice(0, length);
+            }
+
+            public override void Release()
+            {
+                if (_continuousBuffer == null)
```

Comment:
Should this be a volatile read?   What happens if we try to return a buffer that we don't own on L242?

Reply:
[ArrayPool is thread safe](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=net-5.0#thread-safety). How would that happen?

Reply:
I don't believe that it would, under normal circumstances.   Just thinking through what happens if `Dispose` gets manually called multiple times in a tight loop or some other unlikely scenario.